nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/state/server/ZooKeeperStateServerConfigurationsTest.java

```diff
@@ -0,0 +1,211 @@
+        }});
+
+        final ZooKeeperStateServer server = ZooKeeperStateServer.create(secureProps);
+        Assert.assertNotNull(server);
+
+        final QuorumPeerConfig config = server.getQuorumPeerConfig();
+        Assert.assertEquals(secureZooKeeperProps.getProperty("secureClientPort"), String.valueOf(config.getSecureClientPortAddress().getPort()));
+    }
+
+
+    // This shows that a secure NiFi with an insecure ZooKeeper will not have an insecure client address or port:
+    @Test
+    public void testCreateRemovesInsecureClientPort() {
+        Assert.assertNotNull(insecureZooKeeperProps.getProperty("clientPort"));
+        Assert.assertNotEquals(insecureZooKeeperProps.getProperty("clientPort"), "");
+        Assert.assertNull(quorumPeerConfig.getClientPortAddress());
+    }
+
+
+    // This test to shows that an available port is selected when none is specified:
+    @Test
+    public void testCreateWithUnspecifiedSecureClientPort() {
+        final int port = quorumPeerConfig.getSecureClientPortAddress().getPort();
+
+        Assert.assertNotEquals(0, port);
+        Assert.assertNotEquals(1181, port);
+        Assert.assertNotEquals(2181, port);
+        Assert.assertNotEquals(2182, port);
+        Assert.assertTrue(port > ZooKeeperStateServer.MIN_AVAILABLE_PORT);
+    }
+
+    // This test shows that a connection class is set when none is specified (QuorumPeerConfig::parseProperties sets the System property):
+    @Test
+    public void testCreateWithUnspecifiedConnectionClass() {
+        Assert.assertEquals(ZooKeeperStateServer.SERVER_CNXN_FACTORY, System.getProperty(ServerCnxnFactory.ZOOKEEPER_SERVER_CNXN_FACTORY));
+    }
+
+
+    // This test shows that a specified connection class is honored (QuorumPeerConfig::parseProperties sets the System property):
+    @Test
+    public void testCreateWithSpecifiedConnectionClass() throws IOException, QuorumPeerConfig.ConfigException {
+        final NiFiProperties secureProps = NiFiProperties.createBasicNiFiProperties(null, new HashMap<String, String>() {{
+            putAll(SECURE_NIFI_PROPS);
+            put(ZOOKEEPER_PROPERTIES_FILE_KEY, SECURE_ZOOKEEPER_PROPS_URL.getPath());
+        }});
+
+        Assert.assertNotNull(ZooKeeperStateServer.create(secureProps));
+        Assert.assertEquals(secureZooKeeperProps.getProperty("serverCnxnFactory"), System.getProperty(ServerCnxnFactory.ZOOKEEPER_SERVER_CNXN_FACTORY));
+    }
+
+
+    // This test shows that NiFi TLS properties are propagated System properties for ZooKeeper:
+    @Test
+    public void testCreateFromPropagatedNiFiProperties() {
+        ZooKeeperStateServer.ZOOKEEPER_TO_NIFI_PROPERTIES.keySet().forEach((key) -> {
+            Assert.assertEquals( niFiProps.getProperty(ZooKeeperStateServer.ZOOKEEPER_TO_NIFI_PROPERTIES.get(key)),  System.getProperty("zookeeper." + key));
+        });
+    }
+
+
+    private static URL loadResource(String path) {
```

Comment:
I don't think this is necessary. A `new File()` or `Paths.get` relative to "src/test/resources" should work.

Reply:
So simplified.